File: CPP/Sliding Window/Take K of Each Character From Left and Right.cpp
class Solution {
public:
    int takeCharacters(string s, int k) {
        unordered_map<char, int> mp;
        mp['a'] = 0;
        mp['b'] = 0;
        mp['c'] = 0;
        for (auto i : s) {
            mp[i]++;
        }
        if (mp['a'] < k || mp['b'] < k || mp['c'] < k)
            return -1;
        int left = 0;
        unordered_map<char, int> windowFreq;
        int maxWindow=0;
        //Find the longest window that leaves k of each character outside
        for (int right = 0; right < s.size(); right++) {
            windowFreq[s[right]]++;
            // Shrink window if we take too many characters
            while (left<=right && (mp['a']-windowFreq['a'] < k || mp['b']-windowFreq['b'] < k || mp['c']-windowFreq['c'] < k)) {  
                // Shrink the window from the left
                windowFreq[s[left]]--;
                left++;
            }
            maxWindow = max(maxWindow,right-left+1);
        }
        return s.size()-maxWindow;
    }
};

File: Python/Sliding Window/Take K of Each Character From Left and Right.py
class Solution:
    def takeCharacters(self, s: str, k: int) -> int:
        mp = defaultdict(int)
        for i in s:
            mp[i] += 1
        if mp["a"] < k or mp["b"] < k or mp["c"] < k:
            return -1

        left = 0
        windowFreq = defaultdict(int)
        maxWindow = 0

        for right in range(len(s)):
            windowFreq[s[right]] += 1
            while left <= right and (mp["a"] - windowFreq["a"] < k or mp["b"] - windowFreq["b"] < k or mp["c"] - windowFreq["c"] < k):
                # Shrink the window from the left
                windowFreq[s[left]] -= 1
                left += 1

            # Update the maxWindow if the current window is larger
            maxWindow = max(maxWindow, right - left + 1)

        # Return the minimum number of characters to be taken
        return len(s) - maxWindow
